### ISSUE TITLE
Error: bad immediate value for offset (4096)

### ISSUE THREAD

**Issue by nickdesaulniers** [ClangBuiltLinux/linux#490]
https://gist.github.com/nickdesaulniers/37ec31896139b032d915b18eb09fc6cc

Now with asm goto and arm32 allyesconfig, @nathanchance and I observe the above linked failures during assembling. I'm running a creduce now.

blocks #6

**Comment by nickdesaulniers** [comment]
cc @topperc

**Comment by nickdesaulniers** [comment]
cc @smithp35 

so I suspect that there's another step required for inserted global address+offset for arm32, in that if the constant is greater than a number, it's not possible to encode that far of a jump/load/store (reference) in a single instruction?

For example, C reduce produced a 259 line C file, which produces a 1590 line assembly file, which produces the error:

```
$ arm-linux-gnueabihf-as extcon-arizona.s                                
extcon-arizona.s: Assembler messages:
extcon-arizona.s:70: Error: bad immediate value for offset (4096)

$ head extcon-arizona.s -n 70 | tail -n 1
	ldr	r8, .LCPI1_0
```
where `.LCPI1_0` looks far away:

```gas
1343 .LCPI1_0:                                                                       
1344   .long __stack_chk_guard
```

**Comment by nickdesaulniers** [comment]
[extcon-arizona.c.txt](https://github.com/ClangBuiltLinux/linux/files/3209338/extcon-arizona.c.txt)
[extcon-arizona.s.txt](https://github.com/ClangBuiltLinux/linux/files/3209339/extcon-arizona.s.txt)

**Comment by kees** [comment]
ldr immediate maximum is 0xfff (4095). I think large immediate assignment tend to use ldr/add pairs? Maybe something like this?

```
#define IMM12_MASK 0xfff

    ldr r8, .LCPI1_0 & IMM12_MASK
.if .LCPI1_0 > IMM12_MASK
    add r8, r8, .LCPI1_0 & ~IMM12_MASK
.endif
```

Where does llvm make the choice about picking instructions for immediate loading?

I saw something similar when doing randomize_layout on arm, and Ard sorted me out:
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=ffa47aa678cfaa9b88e8a26cfb115b4768325121

**Comment by kees** [comment]
(BTW, extcon-arizona.c doesn't build for me with either gcc nor clang)

**Comment by nickdesaulniers** [comment]
> (BTW, extcon-arizona.c doesn't build for me with either gcc nor clang)

oh! What about the others from the [gist](https://gist.github.com/nickdesaulniers/37ec31896139b032d915b18eb09fc6cc) I posted?

**Comment by smithp35** [comment]
FWIW clang tends to use the register form rather than the immediate form of ldr when the offset is, or could be, larger than 4096
ldr r0, [pc, #offset to constant pool entry with base address]
mov r1, #offset from base address
ldr r2, [r0, r1]

**Comment by nickdesaulniers** [comment]
> FWIW clang tends to use the register form rather than the immediate form of ldr when the offset is, or could be, larger than 4096

Right, so there's probably explicit checks in ISEL(?) and maybe one is missing?

>> (BTW, extcon-arizona.c doesn't build for me with either gcc nor clang)
> What about the others from the gist I posted?

I am very surprised that both compilers fail to compile that.  I'm going to test the rest w/ GCC.

Also, [some failures were no longer reproducible](https://gist.github.com/nickdesaulniers/37ec31896139b032d915b18eb09fc6cc#gistcomment-2925065) w/ `CONFIG_STACKPROTECTOR` disabled.  I don't think that's the root cause though; I think it simply made the space between references shorter.

**Comment by nickdesaulniers** [comment]
Todo: test defconfig+drivers explicitly enabled. Test paring back command
line args.

**Comment by nickdesaulniers** [comment]
Filed upstream bug: https://llvm.org/pr41999 with minimized command line flags.

**Comment by nickdesaulniers** [comment]
results from experiments: https://gist.github.com/nickdesaulniers/37ec31896139b032d915b18eb09fc6cc#gistcomment-2925274.

TL;DR less sure this is asm goto related, but more testing required.

**Comment by nickdesaulniers** [comment]
https://gist.github.com/nickdesaulniers/37ec31896139b032d915b18eb09fc6cc#gistcomment-2925282 testing shows `CONFIG_JUMP_LABEL` (asm goto) is involved it seems.

**Comment by smithp35** [comment]
I've had a quick look and have put a comment in the PR. My suspicion is that the constant island pass that is responsible for placing the constant pool within range of its users is not getting the size of the inline assembly right. The offset in the LDR is only 1 byte out and I think there is some conservative estimate done on the size of inline assembler. Might be able to spend some more time next week to confirm.

**Comment by smithp35** [comment]
My colleague Dave Green thinks that the INLINEASM_BR needs to be added to ARMBaseInstrInfo::getInstSizeInBytes() to prevent the inline assembly size being underestimated. There is a getInstSizeInBytes per Target so it may need to be added to more than just ARM. I think few targets use constant islands, but several may use that function for relaxation and branch optimisation.

**Comment by nickdesaulniers** [comment]
patch: https://reviews.llvm.org/D62400

**Comment by nathanchance** [comment]
With https://reviews.llvm.org/D62400 and https://reviews.llvm.org/D56571:

```
$ git lo HEAD~2..HEAD | cat
9edc4452ed8 D56571
e605b2a5f83 D62400
```

```
clang location: /home/nathan/cbl/tc-build/build/llvm/stage1/bin/clang

clang version 9.0.0 (git://github.com/llvm/llvm-project 9edc4452ed8d0a4ff188f9c7c1938671c4c81a88)

Version: 5.2.0-rc1-cbl-next-20190524+

arm:multi_v7_defconfig | Build successful in 3 minutes and 24 seconds | Boot successful

arm:allyesconfig | Build successful in 14 minutes and 42 seconds

arm64:defconfig | Build successful in 4 minutes and 34 seconds | Boot successful

arm64:allyesconfig | Build successful in 16 minutes and 53 seconds

ppc64le:powernv_defconfig | Build successful in 3 minutes and 35 seconds | Boot successful

x86:defconfig | Build successful in 2 minutes and 25 seconds | Boot successful

x86:allyesconfig | Build successful in 20 minutes and 21 seconds
```

**Comment by nickdesaulniers** [comment]
Fixed: https://reviews.llvm.org/rL361659

**Comment by nickdesaulniers** [comment]
hmmm how did arm build successfully? https://github.com/ClangBuiltLinux/linux/issues/494 were you testing -next?

**Comment by nathanchance** [comment]
Yes but I built mainline successfully with a few patches:

https://github.com/nathanchance/patches/blob/master/linux/error-fixes/0003-ARM-disable-FUNCTION_TRACER-when-building-with-Clang.patch

https://github.com/nathanchance/patches/blob/76d3ce57f6ab93cf215fd6f3e02e23978825f458/linux/error-fixes/0004-ASoC-Intel-Don-t-enable-so-many-boards-for-compile-t.patch

https://github.com/nathanchance/patches/blob/76d3ce57f6ab93cf215fd6f3e02e23978825f458/linux/other-fixes/0005-mlx5-avoid-64-bit-division.patch

And a couple of other ones for ld.lld and objcopy:

https://github.com/nathanchance/patches/blob/76d3ce57f6ab93cf215fd6f3e02e23978825f458/linux/other-fixes/0004-lkdtm-support-llvm-objcopy.patch

https://github.com/nathanchance/patches/blob/76d3ce57f6ab93cf215fd6f3e02e23978825f458/linux/other-fixes/0003-ARM-VDSO-Drop-implicit-common-page-size-linker-flag.patch
